src/ucp/api/ucp.h

```diff
@@ -2885,36 +2900,53 @@ ucs_status_t ucp_mem_advise(ucp_context_h context, ucp_mem_h memh,
  * with the memory handle the application is responsible for sharing the RKEY with
  * the peers that will initiate the access.
  *
- * @param [in]  context       Application @ref ucp_context_h "context" which was
- *                            used to allocate/map the memory.
  * @param [in]  memh          @ref ucp_mem_h "Handle" to memory region.
- * @param [out] rkey_buffer_p Memory buffer allocated by the library.
- *                            The buffer contains packed RKEY.
- * @param [out] size_p        Size (in bytes) of the packed RKEY.
+ * @param [in]  params        Memory handle packing parameters, as defined by
+ *                            @ref ucp_memh_pack_params_t.
+ * @param [out] buffer_p      Memory buffer allocated by the library.
+ *                            The buffer contains the packed memory handle.
+ * @param [out] buffer_size_p Size (in bytes) of the buffer which contains
+ *                            packed memory handle.
  *
  * @return Error code as defined by @ref ucs_status_t
  */
-ucs_status_t ucp_rkey_pack(ucp_context_h context, ucp_mem_h memh,
-                           void **rkey_buffer_p, size_t *size_p);
+ucs_status_t ucp_memh_pack(ucp_mem_h memh, ucp_memh_pack_params_t *params,
```

Comment:
@yosefe since we introduce new API and new buffer object, maybe this an opportunity to break backward compatibility and declare that the new packed object format is not backward compatible.

Reply:
but we also don't want to increase the size of the mkey buffer

Reply:
I think this is the compromise we have to make in order to accommodate wire level protocol support. Probably 1 byte will be enough. I don't think it will have such big impact on openshmem scaling, but you will know better. IMHO, If you at the point that you have scaling problem (like 10^9 endpoints), you have to get rid of keys all together.

Reply:
Well, I guess we can intorduce new APIs both for rkey pack and unpack, and define they return a new binary format.
older applications that use existing APIs will not be affected.
Can use this opportunity to make the binary format extendable.

Reply:
Yep.

Reply:
> Well, I guess we can intorduce new APIs both for rkey pack and unpack, and define they return a new binary format.
> older applications that use existing APIs will not be affected.

updated `ucp_memh_pack` to support RKEY pack compatibility mode.

> Can use this opportunity to make the binary format extendable.

shall we implement new extendable RKEY binary format in this PR?